README
# What is this?

This directory holds Flextime, a program for tracking working hours.

The Flextime daemon query D-Bus or X11 for the time since the last
user input once every minute and stores the result on disk.  The
flextime program can be used to display the times the user has been
active on the computer.  For example:

    $ flextime
    20200810 07:16:50 — 16:10:22  8:53 | 0:27 w/33 Monday
    20200811 13:29:45 — 14:01:18  0:31 | 0:12 w/33 Tuesday
    20200812 07:33:21 — 17:22:56  9:49 | 6:22 w/33 Wednesday
    20200813 07:54:32 — 15:17:12  7:22 | 3:34 w/33 Thursday
    20200814 07:29:37 — 10:38:36  3:08 | 3:08 w/33 Friday

If your working day starts and ends on the computer, this list
effectively represents your working hours.

Just make sure that the Flextime daemon, flextimed, is started every
time you log in.

# Building

`configure` will probe the system for dependencies.  On a Fedora 37
system, I had to install the following packages:

```
dnf install protobuf-c-compiler
dnf install protobuf-c-devel
dnf install libXScrnSaver-devel
dnf install glib2-devel
```

Using the follwing commands to find the packages:

```
dnf provides protoc-c
dnf provides "*/protobuf-c/protobuf-c.h"
dnf provides "*/extensions/scrnsaver.h"
dnf provides "*/gio.h"
```

I do not think that there are any runtime dependencies, except for
X11 and D-Bus (optional.)

# Running the daemon

If you are running a Gnome shell, you can use a desktop file in your
~/.config/autostart/ folder to start the Flextime daemon every time
you log in.  For example:

```
[Desktop Entry]
Type=Application
Encoding=UTF-8
Name=Flextime Daemon
Exec=flextimed
Terminal=false
```

To see logs from the daemon you can use the `journalctl` command:

```
journalctl -r _COMM=flextimed
```

# Create a release

```
git checkout <tag>
make clean dist
# upload tgz to Github release
cp flextime-<version>.tar.gz ~/rpmbuild/SOURCES/
cp rpmbuild/flextime.spec ~/rpmbuild/SPECS/
cd ~/rpmbuild/SPECS/
rpmlint
rpmbuild -ba flextime.spec
```

# Testing

```
$ ./flextime --no-local-data --stop "17:00" --mark "07:00;18:00"
20210826 07:00:00 — 17:00:00 10:00 | 0:00 w/NN Weekday
```

NEWS
# 0.19.0

Measure idle time using D-Bus, if available at compile time.
Otherwise, use X11.

# 0.18.4

Add option --no-continue-on-error (terminate program if a measurement
file cannot be parsed.)

# 0.18.3

Handle failure to read measurement file by printing an error message
and gracefully continuing with the next file, instead of terminating
the program.

# 0.18.2

Close pipe between daemon and print process.

# 0.18.1

Add man pages.

# 0.18.0

Add options --stop and --mark to add measurements from the command
line.  This is useful for answering questions like "what if a stopped
(or started) working at a specific time?"

# 0.17.3

Fix a set of issues where the daemon would hang and stop working.

# 0.17.2

Fix potential buffer overflow

# 0.17.1

Fix build warnings

# 0.17

Fix an issue where running the print command would not properly
display the latest data.  (You had to run the "flextime" command twice
to be sure that you got all the data.)

# 0.16

Add specification file for building an RPM package for Fedora.  The
build process is updated to make it more robust.

The folder that holds the data that the flextime generates is now
created with more restricted access rights (0750).

# 0.15

The daemon process starts as a proper daemon so that flextimed can be
easily started as part of your login shell.

https://github.com/pasce/daemon-skeleton-linux-c

The print command renders usable output, including date, week number,
total time and work time.

Start and stop measurement kinds are written so that we no longer miss
the first and last minute.

# 0.14

Initial version


main.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <X11/extensions/scrnsaver.h>

#include "measurement.pb-c.h"

// https://stackoverflow.com/a/4702411
long get_idle_time() {
  time_t idle_time;
  static XScreenSaverInfo *mit_info;
  Display *display;
  int screen;

  mit_info = XScreenSaverAllocInfo();

  if((display=XOpenDisplay(NULL)) == NULL) {
    return(-1);
  }

  screen = DefaultScreen(display);

  XScreenSaverQueryInfo(display, RootWindow(display,screen), mit_info);

  idle_time = (mit_info->idle) / 1000;

  XFree(mit_info);
  XCloseDisplay(display);

  return idle_time;
}

int main()
{
  char* home = getenv("HOME");

  char topFolderName[1024];
  snprintf(topFolderName, 1024, "%s/.flextime", home);

  char dataFolderName[1024];
  snprintf(dataFolderName, 1024, "%s/.flextime/data", home);

  if (0 == mkdir(topFolderName, 0777)) {
    mkdir(dataFolderName, 0777);
    printf("Created folder %s\n", dataFolderName);
  }

  time_t now = time(NULL);

  long idle = get_idle_time();

  Measurement msg = MEASUREMENT__INIT;
  void *buf;
  unsigned len;

  msg.kind = MEASUREMENT__KIND__MEASUREMENT;
  msg.idle = idle;
  msg.timestamp = (intmax_t)now;

  len = measurement__get_packed_size(&msg);

  buf = malloc(len);
  measurement__pack(&msg,buf);

  char formatTimeBuffer[20];
  strftime(formatTimeBuffer, 20, "%FT%TZ", localtime(&now));

  char fileName[1024];
  snprintf(fileName, 1024, "%s/.flextime/data/%s.bin", home, formatTimeBuffer);
  FILE *file = fopen(fileName, "w+");

  printf("Writing %d bytes of idle %ld to %s\n", len, idle, fileName);

  fwrite(buf, len, 1, file);

  fclose(file);

  free(buf);

  return 0;
}


daemon.c
#include <time.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <X11/extensions/scrnsaver.h>
#include <argp.h>
#include <signal.h>
#include <syslog.h>
#include <fcntl.h>
#include <gio/gio.h>

#include "config.h"
#include "measurement.pb-c.h"

#define PARENT_COUNT 15

static unsigned int foreground = 0;
static unsigned int verbose = 0;
static unsigned int verbose_level = 0;
static unsigned int interval = 60; // seconds, should be called measurement interval

Measurement measurements[PARENT_COUNT]; // the number of measurements within a flush
int idx = 0;

Measurements parent;
void *parent_buf;
unsigned parent_len;

char pipeFileName[1032];

void create_measurements(unsigned int interval, const char* zone) {
  Measurements msgs = MEASUREMENTS__INIT;

  msgs.interval = interval;

  size_t size = strlen(zone);
  msgs.zone = malloc(sizeof(char) * size + 1);

  strcpy(msgs.zone, zone);
  
  msgs.n_measurements = idx;
  msgs.measurements = malloc(sizeof(Measurement) * idx);

  for (int i = 0; i < idx; i++) {
    msgs.measurements[i] = malloc(sizeof(Measurement));
    measurement__init(msgs.measurements[i]);

    msgs.measurements[i]->kind = measurements[i].kind;
    msgs.measurements[i]->idle = measurements[i].idle;
    msgs.measurements[i]->timestamp = measurements[i].timestamp;
  }

  parent_len = measurements__get_packed_size(&msgs);

  parent_buf = malloc(parent_len);
  measurements__pack(&msgs, parent_buf);
  
  parent = msgs;
}

void internal_log(int pri, const char *fmt, ...) {
    va_list vl;
    va_start(vl, fmt);
    if (foreground) {
      vprintf(fmt, vl);
      printf("\n");
      fflush(stdout);
    } else {
      vsyslog(pri, fmt, vl);
    }

    va_end(vl);
}

void flush_pipe() {
  if (verbose) internal_log(LOG_DEBUG, "Writing message to pipe.");

  // Print a message on the named pipe
  int pipe = open(pipeFileName, O_RDWR); // Not O_WRONLY or we will hang if no reader
  ssize_t size = write(pipe, "FLUSHED\n", 8);

  if (size != 8) {
    internal_log(LOG_ERR, "Unexpected write size %ld.\n", size);
  }

  close(pipe);
}

void flush_measurements(int pipe) {
  if (idx == 0) {
    if (pipe) flush_pipe();

    if (verbose) internal_log(LOG_DEBUG, "No measurements to flush.");

    return; // Nothing to flush
  }

  if (verbose) internal_log(LOG_DEBUG, "Flushing measurements at index %d.", idx);

  const time_t now = time(NULL);
  const struct tm *local_now = localtime(&now);

  char formatTimeBuffer[20];
  strftime(formatTimeBuffer, 20, "%FT%TZ", local_now);

  char* home = getenv("HOME");

  char fileName[1024];
  snprintf(fileName, 1024, "%s/.flextime/data/%s.bin", home, formatTimeBuffer);

  if(access(fileName, F_OK ) != -1) {
    // The file already exists.  Since the name of the file is the time in
    // ISO8601 format resolved to seconds, it means that the data is flushed
    // already.
    if (idx == 1) {
      // This can happen, 1/60 of the times we flush within the same second
      internal_log(LOG_WARNING, "One measurement is discarded for file %s.", fileName);
    }

    if (idx > 1) {
      // This can never happen with a measurement interval of one minute
      internal_log(LOG_ERR, "%d measurements are discarded for file %s.", idx, fileName);
    }

    if (idx == 0) {
      // This should never happen
      internal_log(LOG_WARNING, "Zero measurements are discarded for file %s.", fileName);
    }

    return;
  }

  const char* zone = local_now->tm_zone;
  create_measurements(interval, zone);

  FILE *file = fopen(fileName, "w+");

  fwrite(parent_buf, parent_len, 1, file);

  fclose(file);
  if (pipe) flush_pipe();

  if (verbose) internal_log(LOG_DEBUG, "Flushed measurements to file %s.", fileName);
}

void initialize_measurements() {
  for (int i = 0; i < PARENT_COUNT; i++) {
    Measurement msg = MEASUREMENT__INIT;

    parent_len = measurement__get_packed_size(&msg);

    parent_buf = malloc(parent_len);
    measurement__pack(&msg, parent_buf);

    measurements[i] = msg;
  }
}

#if defined USE_GIO && USE_GIO == 1

long
get_idle_time_internal (GDBusProxy *proxy)
{
  guint64 user_idle_time;
  gchar *method = "GetIdletime";
  GError *error = NULL;
  GVariant *ret = NULL;
  
  ret = g_dbus_proxy_call_sync(proxy,
			       method,
			       NULL,
			       G_DBUS_CALL_FLAGS_NONE, -1,
			       NULL, &error);
  if (!ret) {
    g_dbus_error_strip_remote_error (error);
    if (verbose) internal_log(LOG_DEBUG, "GetIdletime returned %s.", error->message);
    g_error_free (error);
    return -1;
  }

  g_variant_get (ret, "(t)", &user_idle_time);
  
  g_variant_unref (ret);
  
  return user_idle_time / 1000;
}

long
get_idle_time ()
{
  GDBusProxy *proxy = NULL;
  gchar *name = "org.gnome.Mutter.IdleMonitor";
  gchar *object_path = "/org/gnome/Mutter/IdleMonitor/Core";
  gchar *interface_name = "org.gnome.Mutter.IdleMonitor";
  /* Create a D-Bus proxy */
  proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
					 G_DBUS_PROXY_FLAGS_NONE,
					 NULL,
					 name,
					 object_path,
					 interface_name,
					 NULL, NULL);
  g_assert (proxy != NULL);
  
  long result = get_idle_time_internal (proxy);
  
  g_object_unref (proxy);
  
  return result;
}


#else /* USE_GIO */

// https://stackoverflow.com/a/4702411
long get_idle_time() {
  time_t idle_time;
  static XScreenSaverInfo *mit_info;
  Display *display;
  int screen;

  mit_info = XScreenSaverAllocInfo();

  if((display=XOpenDisplay(NULL)) == NULL) {
    return(-1);
  }

  screen = DefaultScreen(display);

  XScreenSaverQueryInfo(display, RootWindow(display,screen), mit_info);

  idle_time = (mit_info->idle) / 1000;

  XFree(mit_info);
  XCloseDisplay(display);

  return idle_time;
}
#endif /* USE_GIO */

int finishing = 0;

// sleep(3) can be interrupted by signals
void sleep_with_interrupt(unsigned int seconds) {
    unsigned int left = seconds;
    while ((left > 0) && (!finishing)) {
      left = sleep (left);
    }
}

void create_measurement(Measurement__Kind kind, unsigned int idle) {
  time_t now = time(NULL);

  measurements[idx].kind = kind;
  measurements[idx].idle = idle;
  measurements[idx].timestamp = (intmax_t)now;

  idx++;

  if (verbose) internal_log(LOG_DEBUG, "Created measurement; index now at %d (of %d).", idx, PARENT_COUNT);

  if (idx == PARENT_COUNT) {
    flush_measurements(0);
    idx = 0;
  }
}

void create_measurement_maybe() {
  long idle = get_idle_time();

  // If the idle time is larger than the interval, we do not create a
  // measurement since there is nothing to add to the already created
  // measurements.  The time between measurement is not perfect, so be on
  // the safe side, we compared to a somewhat larger value.
  if (idle > interval * 1.5) {
    if (verbose) internal_log(LOG_DEBUG, "Did not create measurement since idle is too large (%d compared to %f).", idle, interval * 1.5);

    return;
  }

  create_measurement(MEASUREMENT__KIND__MEASUREMENT, idle);
}

void sighandler(int signum) {
  // In general, our approach to flushing data in a signal handler is
  // wrong.  If the main program is in the process of writing data to
  // files, then since the signal handler is also writing to files, the
  // state of the allocated data buffers are simply wrong.
  // 
  // We need to fix that, and avoid using stdio functions from the signal
  // handler.  In fact, only functions listed in signal-safety(7) can be
  // used in a signal handler.
  // 
  // https://stackoverflow.com/a/55179208
  // https://man7.org/linux/man-pages/man7/signal-safety.7.html

   internal_log(LOG_NOTICE, "Caught signal %d, flushing %d measurements.", signum, idx);

   int shutdown = signum != SIGUSR1;

   if (shutdown) {
    if (verbose) internal_log(LOG_DEBUG, "Shutdown measurement created.");
     create_measurement(MEASUREMENT__KIND__STOP, 0);
   }

   if (shutdown) 
    flush_measurements(0);
  else
    flush_measurements(1);

   idx = 0;

   if (shutdown) {
     closelog();

     internal_log (LOG_NOTICE, "Flextime daemon %s terminated.", PACKAGE_VERSION);

     _exit(0);
   }
}

// https://github.com/pasce/daemon-skeleton-linux-c
//
// This might all be wrong-headed, and we should not do it at all:
// https://unix.stackexchange.com/a/177361
static void skeleton_daemon()
{
    if (verbose) internal_log(LOG_DEBUG, "Forking a daemon process.");
    
    pid_t pid;
    
    /* Fork off the parent process */
    pid = fork();
    
    /* An error occurred */
    if (pid < 0)
        exit(EXIT_FAILURE);
    
     /* Success: Let the parent terminate */
    if (pid > 0)
        exit(EXIT_SUCCESS);
    
    /* On success: The child process becomes session leader */
    if (setsid() < 0)
        exit(EXIT_FAILURE);
    
    /* Catch, ignore and handle signals */
    /*TODO: Implement a working signal handler */
    signal(SIGCHLD, SIG_IGN);
    signal(SIGHUP, SIG_IGN);
    
    /* Fork off for the second time*/
    pid = fork();
    
    /* An error occurred */
    if (pid < 0)
        exit(EXIT_FAILURE);
    
    /* Success: Let the parent terminate */
    if (pid > 0)
        exit(EXIT_SUCCESS);
    
    /* Set new file permissions */
    umask(0);
    
    /* Change the working directory to the root directory */
    /* or another appropriated directory */
    int result = chdir("/");

    if (result != 0) {
      internal_log(LOG_ERR, "Unable to change to root directory (%d).", result);
    }
    
    /* Close all open file descriptors */
    int x;
    for (x = sysconf(_SC_OPEN_MAX); x>=0; x--)
    {
        close (x);
    }
    
    /* Open the log file */
    openlog ("flextime", LOG_PID, LOG_DAEMON);
}

const char *argp_program_version = PACKAGE_STRING;
const char *argp_program_bug_address = PACKAGE_BUGREPORT;

static char doc[] = "Flextime -- tracking working hours";

/* A description of the arguments we accept. */
static char args_doc[] = "ARG1 ARG2";

/* The options we understand. */
static struct argp_option options[] = {
  {"foreground", 'f', 0, 0, "Run in the foreground, not as a daemon" },
  {"verbose", 'v', 0, 0, "Print verbose output" },
  { 0 }
};

/* Used by main to communicate with parse_opt. */
struct arguments
{
  char *args[2];                /* arg1 & arg2 */
  int foreground;
  int verbose;
  int verbose_level;
};

/* Parse a single option. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Get the input argument from argp_parse, which we
     know is a pointer to our arguments structure. */
  struct arguments *arguments = state->input;

  switch (key)
    {
    case 'f': 
      arguments->foreground = 1;
      break;

    case 'v':
      arguments->verbose = 1;
      arguments->verbose_level += 1;

      break;

    case ARGP_KEY_ARG:
      if (state->arg_num >= 2)
        /* Too many arguments. */
        argp_usage (state);

      arguments->args[state->arg_num] = arg;

      break;

    case ARGP_KEY_END:
      if (state->arg_num < 0)
        /* Not enough arguments. */
        argp_usage (state);
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

/* Our argp parser. */
static struct argp argp = { options, parse_opt, args_doc, doc };

int main(int argc, char **argv)
{
  struct arguments arguments;

  /* Default values. */
  arguments.foreground = 0;
  arguments.verbose = 0;
  arguments.verbose_level = 0;

  /* Parse our arguments; every option seen by parse_opt will
     be reflected in arguments. */
  argp_parse (&argp, argc, argv, 0, 0, &arguments);

  foreground = arguments.foreground;
  verbose = arguments.verbose;
  verbose_level = arguments.verbose_level;

  char* home = getenv("HOME");

  if (!foreground) {
    skeleton_daemon();
    internal_log(LOG_NOTICE, "Flextime daemon %s started.", PACKAGE_VERSION);
  } else {
    if (verbose) internal_log(LOG_DEBUG, "Running in the foreground.");
  }

  if (verbose_level > 1) {
    internal_log(LOG_INFO, "Verbose level %d specified.", verbose_level);
  }

  if (verbose && (verbose_level > 1)) {
#if defined USE_GIO && USE_GIO == 1
    internal_log(LOG_DEBUG, "Collecting measurements using D-Bus.");
#else
    internal_log(LOG_DEBUG, "Collecting measurements using X11.");
#endif /* USE_GIO */
  }

  char topFolderName[1024];
  snprintf(topFolderName, 1024, "%s/.flextime", home);

  char dataFolderName[1024];
  snprintf(dataFolderName, 1024, "%s/.flextime/data", home);

  if (0 == mkdir(topFolderName, 0750)) {
    mkdir(dataFolderName, 0750);
    internal_log(LOG_NOTICE, "Created folder %s.", dataFolderName);
  }

  snprintf(pipeFileName, 1032, "%s/flushed", topFolderName);
  int pipeFileResult = mkfifo(pipeFileName, S_IRWXU | S_IRGRP);

  if (pipeFileResult != 0) {
    if (errno != EEXIST) {
        // Log a warning and keep going
        internal_log(LOG_WARNING, "Error %d when creating pipe %s (continuing).", errno, pipeFileName);
    }
  }

  signal(SIGINT, sighandler);
  signal(SIGHUP, sighandler);
  signal(SIGQUIT, sighandler);
  signal(SIGABRT, sighandler);
  signal(SIGTSTP, sighandler);
  signal(SIGTERM, sighandler);
  signal(SIGUSR1, sighandler);
  
  initialize_measurements();

  create_measurement(MEASUREMENT__KIND__START, 0);
  sleep_with_interrupt(interval);

  while (1) {
    create_measurement_maybe();
    sleep_with_interrupt(interval);
  }

  if (!foreground) {
    internal_log(LOG_NOTICE, "Flextime daemon %s terminated (should never happen in main).", PACKAGE_VERSION);
    closelog();
  }

  return 1; // We never terminate from main
}


print.c
#define _XOPEN_SOURCE 500 // For getdate(3)
#define _DEFAULT_SOURCE // for scandir(3)
#include <time.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include<sys/dir.h>
#include <unistd.h>
#include <X11/extensions/scrnsaver.h>
#include <argp.h>
#include <signal.h>
#include <time.h>
#include <fcntl.h>

#include "config.h"
#include "measurement.pb-c.h"

#define MAX_MSG_SIZE 1024
#define MAX_PATH 4096 // https://stackoverflow.com/a/9449307

// There will be measurements of kind MEASUREMENT__KIND__STOP in the data
// for any given day, typically.  The intended usage for the --stop
// argument at the command line is to set a hard stop marker so that you
// can answer the question "what if I stopped working at that time?".
// For this we need a new kind of measurment.  The value must be unique,
// so we choose a large value.  At the time of this writing, existing
// measurment kind values are 0-3 (see protobuf definition.)
#define MEASUREMENT__KIND__ARG_STOP 999 // Stop measurement from the command line

static unsigned int fixed = 60 * 10;
static int splitWeeks = 0;
static int read_local_data = 1;
static int continue_on_error = 1;

static size_t
read_buffer (unsigned max_length, uint8_t *out, FILE* file)
{
  size_t cur_len = 0;
  size_t nread;
  while ((nread=fread(out + cur_len, 1, max_length - cur_len, file)) != 0)
  {
    cur_len += nread;
    if (cur_len == max_length)
    {
      fprintf(stderr, "Maximum message length exceeded.\n");
      exit(1);
    }
  }
  return cur_len;
}

static struct timestamp *timestamps = NULL;
static unsigned int ntimestamps = 0;

struct timestamp {
  unsigned int kind;
  unsigned int time;
};

void read_file(FILE* file, char *fileName) {
  Measurements *msgs;

  uint8_t buf[MAX_MSG_SIZE];
  size_t msg_len = read_buffer(MAX_MSG_SIZE, buf, file);

  msgs = measurements__unpack(NULL, msg_len, buf);
  if (msgs == NULL)
  {
    fprintf(stderr, "Error unpacking %s\n", fileName);

    if (continue_on_error) {
      return;
    } else {
      exit(1);
    }
  }

  timestamps = realloc(timestamps, (ntimestamps + msgs->n_measurements) * sizeof(struct timestamp));

  for (int i = 0; i < msgs->n_measurements; i++) {
    Measurement msg = *msgs->measurements[i];

    time_t t = (time_t)msg.timestamp;
    unsigned int idle = msg.idle;

    timestamps[ntimestamps + i].time = t - idle;
    timestamps[ntimestamps + i].kind = msg.kind;
  }

  ntimestamps += msgs->n_measurements;

  measurements__free_unpacked(msgs, NULL);
}

// https://stackoverflow.com/a/744822
int endsWith(const char *str, const char *suffix)
{
    if (!str || !suffix)
        return 0;
    size_t lenstr = strlen(str);
    size_t lensuffix = strlen(suffix);
    if (lensuffix >  lenstr)
        return 0;
    return strncmp(str + lenstr - lensuffix, suffix, lensuffix) == 0;
}

// https://stackoverflow.com/a/6898456
pid_t proc_find(const char* name) 
{
    DIR* dir;
    struct dirent* ent;
    char* endptr;
    char buf[512];

    if (!(dir = opendir("/proc"))) {
        perror("can't open /proc");
        return -1;
    }

    while((ent = readdir(dir)) != NULL) {
        /* if endptr is not a null character, the directory is not
         * entirely numeric, so ignore it */
        long lpid = strtol(ent->d_name, &endptr, 10);
        if (*endptr != '\0') {
            continue;
        }

        /* try to open the cmdline file */
        snprintf(buf, sizeof(buf), "/proc/%ld/cmdline", lpid);
        FILE* fp = fopen(buf, "r");

        if (fp) {
            if (fgets(buf, sizeof(buf), fp) != NULL) {
                /* check the first token in the file, the program name */
                char* first = strtok(buf, " ");
                if (endsWith(first, name)) {
                    fclose(fp);
                    closedir(dir);
                    return (pid_t)lpid;
                }
            }
            fclose(fp);
        }

    }

    closedir(dir);
    return -1;
}

void parseTime(unsigned int seconds, unsigned int *hours, unsigned int *minutes) {
  *hours = (seconds - (seconds % 3600)) / 3600;
  *minutes = ((seconds - *hours * 60 * 60) - (seconds % 60)) / 60;
}

void print_day(time_t start_time, time_t stop_time, unsigned int work) {
  char date[9];
  strftime(date, 9, "%Y%m%d", localtime(&start_time));

  char start[9];
  strftime(start, 9, "%H:%M:%S", localtime(&start_time));

  char stop[9];
  strftime(stop, 9, "%H:%M:%S", localtime(&stop_time));

  char day[14];
  strftime(day, 14, "%A", localtime(&start_time));

  char week[3];
  strftime(week, 3, "%V", localtime(&start_time));

  unsigned int seconds = stop_time - start_time;
  unsigned int hours;
    unsigned int minutes;
  unsigned int work_hours;
  unsigned int work_minutes;

  parseTime(seconds, &hours, &minutes);
  parseTime(work, &work_hours, &work_minutes);

  printf("%s %s — %s %2d:%02d | %d:%02d w/%2s %s\n", date, start, stop, hours, minutes, work_hours, work_minutes, week, day);
}

void print_timestamps() {
  if (ntimestamps == 0) {
    return;
  }

  time_t t = (time_t)timestamps[0].time;
  time_t start_time = (time_t)(timestamps[0].time);
  time_t stop_time = (time_t)(timestamps[0].time);
  unsigned int work = 0;
  unsigned int skip_day = 0;

  char date[9];
  strftime(date, 9, "%Y%m%d", localtime(&t));

  char week[3];
  strftime(week, 3, "%V", localtime(&t));

  for (int i = 0; i < ntimestamps; i++) {
    t = (time_t)(timestamps[i].time);

    char current[9];
    strftime(current, 9, "%Y%m%d", localtime(&t));

    char current_week[3];
    strftime(current_week, 3, "%V", localtime(&t));

    if (strcmp(current, date) != 0) {
      skip_day = 0;
      print_day(start_time, stop_time, work);

      if (strcmp(current_week, week) != 0) {
        if (splitWeeks) {
          printf("\n");
        }

        strcpy(week, current_week);
      }

      start_time = (time_t)(timestamps[i].time);
      stop_time = (time_t)(timestamps[i].time);
      work = 0;

      strcpy(date, current);
    } else {
      // Same day
      if (skip_day) {
        if (i == ntimestamps - 1) {
          // Last measurement AND the last day is a skip
          print_day(start_time, stop_time, work);
        }

        continue;
      }

      stop_time = (time_t)(timestamps[i].time);

      if (i > 0) {
        time_t previous_stop_time = (time_t)(timestamps[i - 1].time);

        if (stop_time - previous_stop_time < fixed) {
          work += stop_time - previous_stop_time;
        }
      }

      if (timestamps[i].kind == MEASUREMENT__KIND__ARG_STOP) {
        skip_day = 1;
      }

      if (i == ntimestamps - 1) {
        // Last measurement
        print_day(start_time, stop_time, work);
      }
    }
  }
}

int intcmp (const void * a, const void * b) {
  return ( ((struct timestamp*)a)->time - ((struct timestamp*)b)->time );
}

void parse_mark_arg(char* mark, int kind) {
  if (mark == NULL) {
    return;
  }

    char *token;
    while ((token = strsep(&mark, ";"))) {
      struct tm *tm_struct;
      tm_struct = getdate(token);

      if (getdate_err != 0) {
        switch(getdate_err) {
        case 1:
          fprintf(stderr, "Unable to use date since the DATEMSK environment variable is not defined, or its value is an empty string.\n");
          break;
        case 2:
          fprintf(stderr, "The template file specified by DATEMSK cannot be opened for reading.\n");
          break;
        case 7:
          fprintf(stderr, "Unable to parse date %s since it is not supported by DATEMSK.\n", mark);
          break;
        case 8:
          fprintf(stderr, "Unable to parse date %s.\n", mark);
          break;
        default:
          fprintf(stderr, "Unable to parse date %s (%d)\n", mark, getdate_err);
          break;
        }

        exit(2);
      } else {
        ntimestamps++;
        timestamps = realloc(timestamps, ntimestamps * sizeof(struct timestamp));
        timestamps[ntimestamps - 1].time = mktime(tm_struct);
        timestamps[ntimestamps - 1].kind = kind;
      }
    }
}

const char *argp_program_version = PACKAGE_STRING;
const char *argp_program_bug_address = PACKAGE_BUGREPORT;

static char doc[] = "Flextime -- tracking working hours";

/* A description of the arguments we accept. */
static char args_doc[] = "ARG1 ARG2";

#define OPT_NO_LOCAL_DATA 1 // --no-local-data
#define OPT_NO_CONTINUE_ON_ERROR 2 // --no-continue-on-error

/* The options we understand. */
static struct argp_option options[] = {
  {"split-week", 'w', 0, 0,  "Split weeks" },
  {"idle", 'i', "minutes", 0, "Idle limit in minutes, default to 10" },
  {"stop", 's', "date", 0, "Mark STOP getdate(3)" },
  {"mark", 'm', "date", 0, "Mark MEASUREMENT getdate(3)" },
  {"no-local-data", OPT_NO_LOCAL_DATA, 0, 0, "Do not read data from local files" },
  {"no-continue-on-error", OPT_NO_CONTINUE_ON_ERROR, 0, 0, "Do not skip files with errors" },
  { 0 }
};

/* Used by main to communicate with parse_opt. */
struct arguments
{
  char *args[2];                /* arg1 & arg2 */
  int splitWeeks;
  unsigned int idle;
  char *stop;
  char *mark;
  int no_local_data;
  int no_continue_on_error;
};

/* Parse a single option. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Get the input argument from argp_parse, which we
     know is a pointer to our arguments structure. */
  struct arguments *arguments = state->input;

  switch (key)
    {
    case 'w': 
      arguments->splitWeeks = 1;
      break;
    case 'i':
      arguments->idle = atoi(arg);
      break;
    case 's':
      arguments->stop = arg;
      break;
    case 'm':
      arguments->mark = arg;
      break;
    case OPT_NO_LOCAL_DATA:
      arguments->no_local_data = 1;
      break;
    case OPT_NO_CONTINUE_ON_ERROR:
      arguments->no_continue_on_error = 1;
      break;


    case ARGP_KEY_ARG:
      if (state->arg_num >= 2)
        /* Too many arguments. */
        argp_usage (state);

      arguments->args[state->arg_num] = arg;

      break;

    case ARGP_KEY_END:
      if (state->arg_num < 0)
        /* Not enough arguments. */
        argp_usage (state);
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

/* Our argp parser. */
static struct argp argp = { options, parse_opt, args_doc, doc };

int main(int argc, char **argv)
{
  struct arguments arguments;

  /* Default values. */
  arguments.splitWeeks = 0;
  arguments.idle = 10;
  arguments.stop = NULL;
  arguments.mark = NULL;
  arguments.no_local_data = 0;
  arguments.no_continue_on_error = 0;

  /* Parse our arguments; every option seen by parse_opt will
     be reflected in arguments. */
  argp_parse (&argp, argc, argv, 0, 0, &arguments);

  fixed = arguments.idle * 60;
  splitWeeks = arguments.splitWeeks;
  read_local_data = !arguments.no_local_data;
  continue_on_error = !arguments.no_continue_on_error;

  char* home = getenv("HOME");

  char dataFolderName[1024];
  snprintf(dataFolderName, 1024, "%s/.flextime/data", home);

  struct stat sb;

  if (stat(dataFolderName, &sb) != 0) {
    printf("Folder %s does not exist.\n", dataFolderName);

    return 1;
  }

  pid_t pid = proc_find("flextimed");

  if (pid >= 0) {
    kill(pid, SIGUSR1);

    char pipeFileName[1024];

    snprintf(pipeFileName, 1024, "%s/.flextime/flushed", home);

    int pipe = open(pipeFileName, O_RDONLY); 
    if (pipe == -1) {
      printf("Pipe file %s open error %d. Scheduling a small sleep instead.", pipeFileName, errno);

      // Wait just a little while for the daemon to flush
      struct timespec ts;
      ts.tv_sec = 1;
      ts.tv_nsec = 0;

      nanosleep(&ts, NULL);
    } else {
      char pipeData[1024];
      ssize_t size = read(pipe, pipeData, sizeof(pipeData));

      if (size == sizeof(pipeData)) {
        printf("Unexpected pipe write size %ld.\n", size);
      }

      // We just read the data, no need to look at it.
      close(pipe);
    }
  }

  struct dirent **namelist;
  if (read_local_data) {
    FILE* fd;

    char path[MAX_PATH];

    int n = scandir(dataFolderName, &namelist, 0, alphasort);
    if (n < 0) {
      perror("scandir");
    }

    while(n--) {
      char *dot = strrchr(namelist[n]->d_name, '.');

      if (dot == NULL || strcmp(dot, ".bin")) {
        continue;
      }

      snprintf(path, MAX_PATH, "%s/%s", dataFolderName, namelist[n]->d_name);

      fd = fopen(path, "r");

      read_file(fd, path);

      fclose(fd);
    }
  }

  parse_mark_arg(arguments.stop, MEASUREMENT__KIND__ARG_STOP);
  parse_mark_arg(arguments.mark, MEASUREMENT__KIND__MEASUREMENT);

  if (ntimestamps > 0){
    qsort(timestamps, ntimestamps, sizeof(struct timestamp), intcmp);

    print_timestamps();
  }

  if (read_local_data) free(namelist);
  if (ntimestamps > 0) free(timestamps);

  return 0;
}
